src/protocol/tx_format/input.md

```diff
@@ -19,20 +19,21 @@ Transaction is invalid if:
-|-----------------------|------------------------------|------------------------------------------------------------------------|
-| `txID`                | `byte[32]`                   | Hash of transaction.                                                   |
-| `outputIndex`         | `uint8`                      | Index of transaction output.                                           |
-| `owner`               | `byte[32]`                   | Owning address or predicate root.                                      |
-| `amount`              | `uint64`                     | Amount of coins.                                                       |
-| `asset_id`            | `byte[32]`                   | Asset ID of the coins.                                                 |
-| `txPointer`           | [TXPointer](./tx_pointer.md) | Points to the TX whose output is being spent.                          |
-| `witnessIndex`        | `uint8`                      | Index of witness that authorizes spending the coin.                    |
-| `maturity`            | `uint32`                     | UTXO being spent must have been created at least this many blocks ago. |
-| `predicateLength`     | `uint16`                     | Length of predicate, in instructions.                                  |
-| `predicateDataLength` | `uint16`                     | Length of predicate input data, in bytes.                              |
-| `predicate`           | `byte[]`                     | Predicate bytecode.                                                    |
-| `predicateData`       | `byte[]`                     | Predicate input data (parameters).                                     |
+| name                  | type                                   | description                                                            |
+|-----------------------|----------------------------------------|------------------------------------------------------------------------|
+| `txID`                | `byte[32]`                             | Hash of transaction or ID of originating message ID.                   |
+| `outputIndex`         | `uint8`                                | Index of transaction output.                                           |
+| `owner`               | `byte[32]`                             | Owning address or predicate root.                                      |
+| `amount`              | `uint64`                               | Amount of coins.                                                       |
+| `asset_id`            | `byte[32]`                             | Asset ID of the coins.                                                 |
+| `txPointer`           | [TXPointer](./tx_pointer.md)           | Points to the TX whose output is being spent.                          |
+| `messagePointer`      | [MessagePointer](./message_pointer.md) | Points to the message being spent.                                     |
```

Comment:
I don't think adding message fields to coins is a clean solution, as fields like tx and outputIndex are no longer relevant. We also lose track of the message ID which is important for verifying in the consensus rules.

I'd prefer to have a new distinct type representing Coins that come from messages, which don't have conflicting identifier fields (e.g. `MessageCoin` or `DepositCoin`)